### DISCUSSION TITLE
A new XYZ-grid contrib

### DISCUSSION THREAD

**Discussion by Griatch** [evennia/evennia#2473] [Evennia Engine Development]
After a bunch of work, I yesterday merged the first version of the `XYZGrid` contrib to develop branch (what will become Evennia 1.0). This is an optional system for placing Evennia rooms on a coordinate system. It includes pathfinding (auto-walk shortest path to target) and map-display in-game. 
Check out the docs here: https://www.evennia.com/docs/1.0-dev/Contribs/XYZGrid.html

[![asciicast](https://asciinema.org/a/Zz36JuVAiPF0fSUR09Ii7lcxc.svg)](https://asciinema.org/a/Zz36JuVAiPF0fSUR09Ii7lcxc)

.
Griatch

**Comment by pruuk** [comment]
This is amazing and is going into my game in a modified form. Thanks!

**Reply by Griatch** [reply]
Cool, can you elaborate on what the 'modified form' will be?

**Reply by pruuk** [reply]
TL:DR; I'm not sure yet.

Previously, I used Z coordinate for rooms above and below (air, underground, and water environments). I'm also trying to wrap my head around how this will work with the tracking related talents I'll have. The good news is that I am starting over from scratch again (5th, time exceedingly efficient at it, etc), so I don't have to unwind anything, just pull in pieces of what I did before.

**Reply by Griatch** [reply]
Is see. Well, just ask if you have any questions, of course. :smiley:

**Reply by pruuk** [reply]
I got pretty excited reading through the docs on this, but didn't notice this is in the develop branch. I tried pulling it in my adding the folder into my MUD, but it seems like there are a number of other dependencies I missed. Is there a guide somewhere about bringing in stuff from the develop branch?

**Reply by Griatch** [reply]
No, the develop branch content cannot easily be moved/mixed with master branch content.

**Comment by ChrisLR** [comment]
How does changing the typeclass work?
I've tried to create a new prototype with a typeclass and the created rooms would be the standard XYZGrid Room.
I tried to use the setting XYZROOM_PROTOTYPE_OVERRIDE and changing the typeclasses in the XYMAP_DATA but then I get errors like

evennia.contrib.xyzgrid.xyzroom.DoesNotExist: XYZRoom matching query xyz=(2,0,hospital), db_typeclass_path__in=['evennia.contrib.xyzgrid.xyzroom.XYZRoom'] does not exist.

The real reason I'm doing this is to be able to populate these rooms automatically, as far as I know the prototypes themselves don't allow that, maybe I'm wrong?

**Reply by Griatch** [reply]
Changing the typeclass in the prototype will, well, use that prototype for your typeclass. However, for xyz-coordinates to work, your overriding typeclass must be a child of `XYZRoom`, maybe that's what you are missing? 

If by 'populating a room' means filling it with objects, then that's not what a prototype is for. A prototype is a structured way to compactify the arguments used to create an object - there is no real context of a 'contents' of objects since this is just objects whose location is set to you.

All the XYZGrid is dealing with is the topology of the grid. You can set descs etc since these are Attributes, but beyond that you are really expected to either step in the grid and flesh things out manually, or add some additional scripting in your typeclasses (which it sounds like are the approach you are taking).

**Reply by ChrisLR** [reply]
Yeah, my new typeclass does inherit from XYZRoom, the rooms are being created right it seems it fails when creating the exits..
Anyway I'll do some more testing and see.
Thanks for the help and for the awesome contrib :)

**Reply by ChrisLR** [reply]
It seems the DefaultRoom is not taken into effect when spawning the XYZGrid rooms, or maybe I goofed again :P
I'm using the default XYZRoom and used the settings 
BASE_ROOM_TYPECLASS = "typeclasses.rooms.Room"

The overloaded code is not taken into effect in the XYZGrid Rooms.
If I go ingame in a XYZRoom, use the "@py" command and print out self.location.is_typeclass("typeclasses.rooms.Room") it return False.

If I dig out any room out of limbo it is of the proper default type.

Do I need to subclass the XYZRoom in order to use my own default room?

**Reply by Griatch** [reply]
Hm, this may be my goof. I can't check right now, but I think the XYZRoom may be inheriting directly from `DefaultRoom` rather than from `BASE_ROOM_TYPECLASS`, meaning your changes will be out of the loop. I'll check this.

**Reply by Griatch** [reply]
@ChrisLR I looked into this again and while it's true that `XYZRoom/Exit` doesn't inherit from the typeclasses set by `BASE_ROOM/EXIT_TYPECLASS` (but directly from `DefaultRoom/Exit`), I don't think this is really a bug. You will *never* spawn regular rooms with xyzgrid - regular rooms don't have what's needed. 

Such rooms must ALWAYS be children of `XYZRoom`/`XYZExit`, so setting them to inherit from `BASE_ROOM/EXIT_TYPECLASS` is backwards. The `BASE_ROOM/EXIT_TYPECLASS` are still relevant - for rooms that are not on the xyz-grid. 

If you want the spawner to spawn rooms of your own design, you must make your own Room-child of XYZRoom (this you can inherit from your base typeclass via multi-inheritance if you like) and then you set the `typeclass` of each coordinate's prototype to use your custom typeclass. 

To avoid rewriting a lot of the same for a lot of prototypes you can follow the example in the `xyzgrid`-contrib `example.py`, where each prototype gets updated with these values automatically (it's just a dict after all). You can also use the `XYZROOM/EXIT_PARENT_PROTOTYPE_OVERRIDE` settings.

**Reply by ChrisLR** [reply]
Can't get it to work, when the spawner tries to create exits it doesn't find any rooms and the new typeclass is not in the db_typepath_class list. I will open an issue with more details.

**Comment by ChrisLR** [comment]
MapTransitions with UpDown Exits don't seem to play nice together.
From the docs I thought
```
>
u
#
```
Would bind the > MapTransition to the Up Direction of the node but it doesn't work, saying the node to the S does not fulfill the criteria.
Is there something I'm forgetting?

**Reply by Griatch** [reply]
@ChrisLR Hm, I tested this quickly by making this local change to the 'small cave' example map: 

Changed 
```
T-#   
```
to
```
T
u
#-# 
```

This worked fine, and I walk 'up' to go to the other map. 

The transitional node must have exactly one link connecting to it, do you have other things on your map (not shown in your example) that also connect to your transition. Also, `>` is by default a one-way west-east link, are you sure to you remapped it to a transitional link in the map legend?

**Reply by ChrisLR** [reply]
You're right, I thought the LEGEND could be defined at module level, placing it in the XYMAP worked.
I got confused at this example
https://www.evennia.com/docs/1.0-dev/Contribs/XYZGrid.html#map-transition-nodes
Thanks!

**Reply by Griatch** [reply]
Ah, I see what you mean. These examples are meant to show off parts of the setup exemplified earlier on the page, but I can see that `LEGEND` looks like its own thing in this example. I'll try to clarify that.

**Reply by Griatch** [reply]
I updated the doc to clearify the LEGEND bit more. :)

**Comment by Henddher** [comment]
Hi!
Is it possible to specify the `typeclass` of a grid room/exit directly within the `"prototypes"` dict of a map?

https://github.com/evennia/evennia/tree/develop/evennia/contrib/grid/xyzgrid#defining-an-xymap

It seems like one has to override `"xyz_room"` and `"xyz_exit"` prototypes using `XYZROOM_PROTOTYPE_OVERRIDE` and `XYZEXIT_PROTOTYPE_OVERRIDE`?

https://github.com/evennia/evennia/tree/develop/evennia/contrib/grid/xyzgrid#prototypes

**Reply by InspectorCaracal** [reply]
Not at all, you can use whatever you like.

Here's an abridged example taken from a map I built a while back which shows how to specify your own typeclass, among other things:

```py
ROOM_PARENT = {
	"key": "An empty room",
	"prototype_parent": "xyz_room",
	"typeclass": "world.map.typeclasses.CustomXYZRoom",
	"desc": "An empty room.",
}

PROTOTYPES = {
	(39, 58): {
		"key": "The Well",
		"desc": "A central road running from the city gates through the Well district. Something something gathering places.",
		"prototype_parent": ROOM_PARENT,
		"typeclass": "world.map.typeclasses.StreetRoom",
		"tags": [ ("outside", "system") ]
	},
# ...
}

XYMAP_DATA_GROUND = {
	"zcoord": "the city",
	"map": CITY_MAP,
	"legend": LEGEND,
	"prototypes": PROTOTYPES,
}
```

So, you can create your own prototype that specifies your own typeclass and then use that as a prototype parent. Or, you can specify the typeclass directly into your prototypes dict.

**Reply by InspectorCaracal** [reply]
The XYZ contrib uses the built-in evennia prototypes system, so you might find some more useful information in the documentation on them directly: https://www.evennia.com/docs/1.0-dev/Components/Prototypes.html
